### PR TITLE
Switch to first allowed type if default type is not allowed in element group

### PR DESCRIPTION
**Pull request by dennisbohn**
Currently, the type field of a content element is set to "article" by default, which leads to an unknown option being displayed in the drop-down list of the type if you create it in an element group where the type "article" is not contained in "allowedTypes".

```
#[AsContentElement(category: 'custom', template:'content_element/element_group', nestedFragments: ['allowedTypes' => ['image']])]
class ImageGroupController extends AbstractContentElementController
{
    protected function getResponse(FragmentTemplate $template, ContentModel $model, Request $request): Response
    {
        return $template->getResponse();
    }
}
```

image

In this case, the added callback function automatically changes the value to the first supported type.

### CODE REVIEW

**Review by fritzmg — CHANGES_REQUESTED** [review]
Agreed. However, new DataContainer callbacks should always be a proper service.

**Comment by dennisbohn** [comment]
Thanks for the review. The code is now a service. It seems that a test case is still expected here. Unfortunately, I currently have no idea how to build one.

File: core-bundle/src/EventListener/DataContainer/CteAllowedTypeListener.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ) {
+    }
+
+    /**
+     * Switch to first allowed type if default type is not allowed (see #7100).
+     */
+    #[AsCallback(table: 'tl_content', target: 'config.oncreate')]
+    public function setAllowedType(string $strTable, int $insertID, array $values, DataContainer $dc): void
+    {
+        $objCte = ContentModel::findById($insertID);
```
Comment (fritzmg) [inline core-bundle/src/EventListener/DataContainer/CteAllowedTypeListener.php]:
```suggestion
        $element = ContentModel::findById($insertID);
```

We only use the type prefix in variable names in legacy code.

**Comment by aschempp** [comment]
This should already be implemented. If I remember correctly, `tl_content` has a method for that. And my permission PR for tl_content rewrites that as well if I remeber correctly 🤔

**Comment by dennisbohn** [comment]
> This should already be implemented. If I remember correctly, `tl_content` has a method for that. And my permission PR for tl_content rewrites that as well if I remeber correctly 🤔

The select box for the type is filtered in the "getContentElements" function, but the default value "text" remains, regardless of whether "text" was specified as allowedType or not.

File: core-bundle/src/EventListener/DataContainer/CteAllowedTypeListener.php (hunk 1 above)
Comment (dennisbohn) [reply core-bundle/src/EventListener/DataContainer/CteAllowedTypeListener.php]: done.

**Comment by aschempp** [comment]
It still is present in 5.3: https://github.com/contao/contao/blob/5.3/core-bundle/contao/dca/tl_content.php#L1089

**Comment by fritzmg** [comment]
> It still is present in 5.3: https://github.com/contao/contao/blob/5.3/core-bundle/contao/dca/tl_content.php#L1089

That has no effect though as the SQL default is `text`. A DCA default cannot override a database level default.

**Comment by dennisbohn** [comment]
The database default takes effect if no value has been set. If a value is set by Contao, the default value in MySQL should no longer apply.

If I understand correctly, the code here only applies to global type restrictions of the user, not to restrictions in an element group.

Restrictions at user level would indeed still have to be taken into account in my PR. Currently, if "text" is not included in the allowedTypes, the first value from allowedTypes is set. Even if the user has no authorization here.

The question is, what should happen if only types for which the user has no authorization are allowed in an element group?

**Comment by dennisbohn** [comment]
Below you will find a different approach that does not use a service for oncreate_callback, but adds code to the filterContentElements function.

From line 1090 there is a new if-Block for element groups.

https://github.com/dennisbohn/contao/blob/1f4e92de129a68d1110a05f27fbdc68e64f40ef3/core-bundle/contao/dca/tl_content.php#L1073-L1112

Commit: https://github.com/dennisbohn/contao/commit/1f4e92de129a68d1110a05f27fbdc68e64f40ef3

**Comment by dennisbohn** [comment]
Below you find a different approach that does not use a service for oncreate_callback, but adds code to the filterContentElements function.

In this solution the edit and create buttons are deactivated if the types are restricted and the user does not have the corresponding authorization.

From line 1089 there is a new if-Block for element groups.

https://github.com/dennisbohn/contao/blob/3bd7c0de8ca6703812848ccec566d860e0c7381b/core-bundle/contao/dca/tl_content.php#L1089-L1103

Commit: https://github.com/dennisbohn/contao/commit/3bd7c0de8ca6703812848ccec566d860e0c7381b

**Comment by leofeyer** [comment]
@dennisbohn As discussed in the Contao call, https://github.com/dennisbohn/contao/commit/3bd7c0de8ca6703812848ccec566d860e0c7381b is the correct fix. Can you adjust your pull request accordingly please?

**Comment by aschempp** [comment]
> https://github.com/dennisbohn/contao/blob/3bd7c0de8ca6703812848ccec566d860e0c7381b/core-bundle/contao/dca/tl_content.php#L1089-L1103

@dennisbohn are you sure this works? I would assume `$dc->getCurrentRecord()` would return the _current_ element in `edit` action, and if that does not support nested types it would close the DC?

**Comment by dennisbohn** [comment]
Ok, I will check and change this. Unfortunately, the pull request was automatically closed when I undid the changes. I create a new pull request.

**Comment by dennisbohn** [comment]
Ok, I will check and change this.

**Comment by dennisbohn** [comment]
After checking the code, I have to admit that @fritzmg's comment is justified.

Although the filterContentElements method is also executed when a content element is created, it is currently somewhat confusing that parts of the code are then responsible for the creation process, while others are responsible for editing.

The fact that the ID of the parent element is still in the URL during creation but the ID of the content element during editing has meant that I had to outsource the logic to another method to keep the code clear.

To keep everything clear, I would therefore use a check within a proper service, as recommended by @fritzmg.

**Comment by dennisbohn** [comment]
Unfortunately, I broke this pull request. The new one can be found here. https://github.com/contao/contao/pull/7196

**Pull request closed without merging**